src/Patterns/Patterns.php

```diff
@@ -0,0 +1,193 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Patterns;
+
+use Automattic\WooCommerce\Blocks\Domain\Package;
+
+/**
+ * Patterns class
+ */
+class Patterns {
+	/**
+	 * Path to the patterns directory.
+	 *
+	 * @var string $patterns_path
+	 */
+	private $patterns_path;
+
+	/**
+	 * Constructor for class
+	 *
+	 * @param Package $package An instance of Package.
+	 */
+	public function __construct( Package $package ) {
+		$this->patterns_path = $package->get_path( 'patterns' );
+
+		add_action( 'init', array( $this, 'register_block_patterns' ) );
+	}
+
+	/**
+	 * Register patterns under the `./patterns/` directory.
+	 * Each pattern is defined as a PHP file and defines its metadata using plugin-style headers.
+	 * The minimum required definition is:
+	 *
+	 *     /**
+	 *      * Title: My Pattern
+	 *      * Slug: my-theme/my-pattern
+	 *      *
+	 *
+	 * The output of the PHP source corresponds to the content of the pattern, e.g.:
+	 *
+	 *     <main><p><?php echo "Hello"; ?></p></main>
+	 *
+	 * Other settable fields include:
+	 *
+	 *   - Description
+	 *   - Viewport Width
+	 *   - Categories       (comma-separated values)
+	 *   - Keywords         (comma-separated values)
+	 *   - Block Types      (comma-separated values)
+	 *   - Inserter         (yes/no)
+	 *
+	 * @since 6.0.0
+	 */
+	public function register_block_patterns() {
+		if ( ! class_exists( 'WP_Block_Patterns_Registry' ) ) {
+			return;
+		}
+
+		$default_headers = array(
+			'title'         => 'Title',
+			'slug'          => 'Slug',
+			'description'   => 'Description',
+			'viewportWidth' => 'Viewport Width',
+			'categories'    => 'Categories',
+			'keywords'      => 'Keywords',
+			'blockTypes'    => 'Block Types',
+			'inserter'      => 'Inserter',
+		);
+
+		if ( file_exists( $this->patterns_path ) ) {
+			$files = glob( $this->patterns_path . '/*.php' );
+			if ( $files ) {
+				foreach ( $files as $file ) {
+					$pattern_data = get_file_data( $file, $default_headers );
+
+					if ( empty( $pattern_data['slug'] ) ) {
+						_doing_it_wrong(
+							'register_block_patterns',
+							esc_html(
+								sprintf(
+								/* translators: %s: file name. */
+									__( 'Could not register file "%s" as a block pattern ("Slug" field missing)', 'woo-gutenberg-products-block' ),
+									$file
+								)
+							),
+							'6.0.0'
+						);
+						continue;
+					}
+
+					if ( ! preg_match( '/^[A-z0-9\/_-]+$/', $pattern_data['slug'] ) ) {
```

Comment:
Can we comment this regex, or move it to a named variable so its more descriptive?